proof/refine/ARM/Finalise_R.thy

```diff
@@ -2326,6 +2326,295 @@ lemma replyClear_invs'[wp]:
+  apply (drule_tac k=ko in ko_at_valid_objs'; clarsimp simp: projectKOs valid_obj'_def
+                 valid_sched_context'_def valid_sched_context_size'_def objBits_def objBitsKO_def)
+  apply (drule_tac k=koa in ko_at_valid_objs'; clarsimp simp: projectKOs valid_obj'_def
+                 valid_sched_context'_def valid_sched_context_size'_def objBits_def objBitsKO_def)
+  apply (clarsimp simp: valid_reply'_def)
+  done
+
+lemma updateReply_list_refs_of_replies':
+  "\<lbrace>\<lambda>s. \<forall>r. ko_at' r replyPtr s \<longrightarrow>
+              P ((list_refs_of_replies' s)(replyPtr := list_refs_of_reply' (f r)))\<rbrace>
+   updateReply replyPtr f
+   \<lbrace>\<lambda>rv s. P (list_refs_of_replies' s)\<rbrace>"
+  by (wpsimp simp: updateReply_def)
+
+lemma replyPop_list_refs_of_replies'[wp]:
+  "\<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s) \<and> obj_at' (\<lambda>reply. replyNext reply \<noteq> None) replyPtr s\<rbrace>
+   replyPop replyPtr tcbPtr
+   \<lbrace>\<lambda>_ s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  unfolding replyPop_def
+  apply (rule hoare_seq_ext[OF _ get_reply_sp'])
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)+
+  apply (rule hoare_seq_ext)
+   apply (wpsimp wp: cleanReply_list_refs_of_replies')
+  apply (rule hoare_when_cases)
+   apply (clarsimp simp: obj_at'_def)
+  apply (rule hoare_seq_ext[OF _ assert_sp])
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp simp: comp_def\<close>, simp?)+
+  apply (rule hoare_seq_ext)
+   apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)
+   apply (wpsimp wp: hoare_when_cases)
+  apply (rule hoare_when_cases[rotated])
+   apply (wp updateReply_list_refs_of_replies' hoare_vcg_all_lift hoare_drop_imp)
+   apply clarsimp
+   apply (rename_tac prevReplyPtr prevReply)
+   apply (frule_tac reply=prevReply in ko_at'_replies_of')
+   apply (frule (3) sym_refs_replyNext)
+   apply (clarsimp,
+          erule delta_sym_refs
+          ; clarsimp simp: isHead_def obj_at'_def projectKOs list_refs_of_reply'_def
+                           list_refs_of_replies'_def opt_map_def get_refs_def
+                    split: if_splits option.splits)+
+  done
+
+lemma replyPop_iflive:
+  "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s \<and> valid_objs' s \<and>
+        ex_nonz_cap_to' tcbPtr s\<rbrace>
+   replyPop replyPtr tcbPtr
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  unfolding replyPop_def
+  supply if_split[split del]
+  apply (wpsimp wp: schedContextDonate_if_live_then_nonz_cap' hoare_vcg_if_lift2 hoare_vcg_conj_lift
+                    hoare_vcg_imp_lift' threadGet_const)
+                  apply (wp updateReply_obj_at'_only_st_qd_ft)
+                 apply (wpsimp wp: updateReply_iflive'_strong replyNext_update_valid_objs')
+                apply (wpsimp wp: replyNext_update_valid_objs' hoare_vcg_imp_lift')+
+  sorry
+
+lemma updateReply_obj_at'_sc_state:
```

Comment:
Is this lemma name what it is supposed to be?

Reply:
Oh I haven't updated the name for this one after creating it. Maybe `updateReply_obj_at'_sc_ksIdleThread`?

Reply:
yes, that sounds good!